Add an in-memory-preloading loader for Jinja imports

depoll opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

#88624 was locked, so opening a new PR with a draft of a new approach based on @balloob's feedback.  Happy to revise based on feedback.  Entirely possible I've chosen the wrong point for initialization, etc.

Breaking out part of https://github.com/home-assistant/core/pull/88575

Enables [importing/including](https://jinja.palletsprojects.com/en/3.1.x/templates/#include) shared jinja code/macros from a custom_jinja folder in your config folder.

All of this enables the following code:

In config/custom_jinja/shared.jinja:

```jinja2
{% macro some_macro() -%}
A macro
{%- endmacro %}
{%- set some_constant = "A constant" %}
```

Elsewhere:

```
{%- import "shared.jinja" as shared %}
{{ shared.some_constant }}
{{ shared.some_macro() }}
```

With the following output:

```
A constant
A macro
```

I'm happy to update the documentation as well but wanted to first clear that this would be a welcome change.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/26436

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

depoll committed:
Adds a loader to enable jinja imports.

depoll committed:
Switch to in-memory

depoll committed:
Move loading custom_jinja off of the event loop

depoll committed:
Raise TemplateNotFound if template doesn't exist

depoll committed:
Merge branch 'dev' into jinja-import

depoll committed:
Fix docstring

depoll committed:
Adds a service to reload custom jinja

balloob commented on tests/common.py:
```diff
@@ -255,6 +256,7 @@ def async_create_task(coroutine):
                 device_registry.async_load(hass),
                 entity_registry.async_load(hass),
                 issue_registry.async_load(hass),
+                template.async_materialize_hass_loader(hass),
```
This adds I/O to every mock hass object in every test, while it's never used. We shouldn't load it here. If needed, just insert an empty HassLoader object.

depoll replied:
Whoops -- 100% right. Will fix.

depoll replied:
Done.

depoll committed:
Remove IO from test setup

emontnemery reviewed (changes requested):
Please add or improve tests so we have full coverage of the added code.

depoll committed:
Improve coverage and small refactor

depoll commented:
> Please add or improve tests so we have full coverage of the added code.

Fixed -- sorry, a change I had made accidentally disabled a test.  Added a bit more as well.

balloob commented on homeassistant/components/homeassistant/__init__.py:
```diff
@@ -305,21 +317,32 @@ async def async_handle_reload_all(call: ha.ServiceCall) -> None:
+            ]
+            + [
+                hass.services.async_call(
+                    ha.DOMAIN,
+                    SERVICE_RELOAD_CUSTOM_JINJA,
```
you should be able to include this in the previous list comprehension.

balloob commented on homeassistant/components/homeassistant/__init__.py:
```diff
@@ -305,21 +317,32 @@ async def async_handle_reload_all(call: ha.ServiceCall) -> None:
+                    domain, service, context=call.context, blocking=True
+                )
+                for domain, service in {
+                    ha.DOMAIN: SERVICE_RELOAD_CORE_CONFIG,
```
```suggestion
                    ha.DOMAIN: SERVICE_RELOAD_CORE_CONFIG,
                    ha.DOMAIN: SERVICE_RELOAD_CUSTOM_JINJA,
```

depoll replied:
That results in a duplicate key in the dictionary.  Will switch to a list of tuples here.

balloob commented on homeassistant/components/homeassistant/services.yaml:
```diff
@@ -59,6 +59,12 @@ update_entity:
     entity: {}
 
+reload_custom_jinja:
+  name: Reload custom Jinja2 code
```
Is "code" the right word to use?

depoll replied:
The Jinja documentation calls them "templates", so I'll switch to that terminology. https://jinja.palletsprojects.com/en/3.1.x/templates/#import-context-behavior

balloob commented on homeassistant/helpers/template.py:
```diff
@@ -2048,6 +2050,61 @@ def __bool__(self) -> bool:
+        if item.is_file() and item.stat().st_size < 1024 * 1024
+    ]
+    for file in all_files:
+        content = file.open().read()
```
this won't close the file.

balloob replied:
```suggestion
        content = file.read_text()
```

balloob commented on homeassistant/helpers/template.py:
```diff
@@ -2048,6 +2050,61 @@ def __bool__(self) -> bool:
+
+
+def _get_hass_loader(hass: HomeAssistant) -> HassLoader:
+    if _HASS_LOADER not in hass.data:
+        hass.data[_HASS_LOADER] = HassLoader({})
```
Use the `singleton` decorator from `helpers/singleton.py` that will do this automatically for you.

balloob commented on homeassistant/helpers/template.py:
```diff
@@ -2063,6 +2120,8 @@ def __init__(self, hass, limited=False, strict=False):
         self.template_cache: weakref.WeakValueDictionary[
             str | jinja2.nodes.Template, CodeType | str | None
         ] = weakref.WeakValueDictionary()
+        if hass is not None:
```
when is `hass` None ?

depoll replied:
There is also a check for `hass is None` below, and IIRC there were at least testing scenarios that had None as a value -- this was initially added to appease them.  For now, I'm going to move it down to after the later check and get rid of the extra condition.

depoll commented on homeassistant/helpers/template.py:
```diff
@@ -2048,6 +2050,61 @@ def __bool__(self) -> bool:
+    jinja_path = hass.config.path("custom_jinja")
+    all_files = [
+        item
+        for item in pathlib.Path(jinja_path).rglob("*")
```
Let's use .jinja, which is what their docs recommend: https://jinja.palletsprojects.com/en/3.0.x/templates/#template-file-extension

depoll committed:
Incorporate feedback and use .jinja extension

depoll committed:
Check the loaded sources in test.

depoll commented:
Added a link to a documentation PR.

emontnemery commented on homeassistant/helpers/template.py:
```diff
@@ -2048,6 +2051,60 @@ def __bool__(self) -> bool:
         return super().__bool__()
 
 
+async def async_materialize_hass_loader(hass: HomeAssistant) -> None:
```
The docstring says this will load custom jinja, can we make the function name more distinct?
```suggestion
async def async_load_custom_jinja(hass: HomeAssistant) -> None:
```

emontnemery commented on homeassistant/helpers/template.py:
```diff
@@ -2048,6 +2051,60 @@ def __bool__(self) -> bool:
+    return await hass.async_add_executor_job(_materialize_hass_loader, hass)
+
+
+def _materialize_hass_loader(hass: HomeAssistant) -> None:
```
Rename this too

emontnemery commented on homeassistant/helpers/template.py:
```diff
@@ -2048,6 +2051,60 @@ def __bool__(self) -> bool:
+    all_files = [
+        item
+        for item in pathlib.Path(jinja_path).rglob("*.jinja")
+        if item.is_file() and item.stat().st_size < 5 * 1024 * 1024
```
Let's make the magic number a constant and also document the limitation

emontnemery commented on homeassistant/helpers/template.py:
```diff
@@ -2151,6 +2208,9 @@ def __init__(self, hass, limited=False, strict=False):
         if hass is None:
             return
 
+        # If hass is available, attach its loader to enable imports.
```
The comment suggests something happens only if hass is available. What if it isn't?

depoll committed:
Incorporate PR feedback.

emontnemery commented on homeassistant/helpers/template.py:
```diff
@@ -2151,6 +2210,9 @@ def __init__(self, hass, limited=False, strict=False):
         if hass is None:
             return
 
+        # When hass is available, attach its loader to enable imports.
```
```suggestion
        # This environment has access to hass, attach its loader to enable imports.
```

depoll committed:
Update homeassistant/helpers/template.py

Co-authored-by: Erik Montnemery <erik@montnemery.com>

depoll committed:
Merge branch 'dev' into jinja-import

emontnemery reviewed (approved):
This is great, thanks @depoll 👍
